submissions/xsorter/port-sniffer/program.js

```diff
@@ -0,0 +1,95 @@
+const net = require('net');
+const dns = require('dns');
+const args = require('minimist')(process.argv.slice(2));
+const range = findPortsRange();
+
+const ports = {
+  firstPort: +range[0],
+  lastPort: +range[1] ? +range[1] : +range[0],
+  openedPorts: []
+};
+
+const openedPortCheck = (host, port, checkCallback) => {
+  const socket = net.createConnection(port, host);
+  const time = socket.setTimeout(300, () => {
+    socket.destroy();
+    checkCallback(false);
+  });
+
+  socket.on('connect', () => {
+    clearTimeout(time);
+    socket.destroy();
+    process.stdout.write('.');
+    ports.openedPorts.push(port);
+    checkCallback(true);
+  });
+  socket.on('error', function () {
+    clearTimeout(time);
+    checkCallback(false);
+  });
+};
+
+const showResult = host => {
+  if (args.help) {
+    process.stdout.write(messages().help);
+    process.exit(1);
+  }
+  if (!args.host) {
+    process.stdout.write(messages().noHost);
+    process.exit(1);
+  } else {
+    openedPortCheck(host, ports.firstPort, function nextIteration () {
```

Comment:
This recursion is not effective due to huge memory usage, slowness, and other reasons.

I propose to you rewrite this to one of the ways:
1. Async Iterations (can be paralleled like 10+ in one batch iterations, looks how https://caolan.github.io/async/v3/docs.html#mapLimit works)
2. Generator (if use 12+ nodejs -> async generator)
3. If you want recursion, let it be optimized, like tails recursion

Reply:
Thanks, i changed recursion with mapLimit